src/Microsoft.ComponentDetection.Detectors/pnpm/PnpmComponentDetector.cs

```diff
@@ -104,14 +125,97 @@ private void RecordDependencyGraphFromFile(PnpmYaml yaml, ISingleFileComponentRe
+
+        // Create a component for every package referenced in the lock file.
+        // This includes all directly and transitively referenced dependencies.
+        foreach (var (pnpmDependencyPath, package) in yaml.packages ?? Enumerable.Empty<KeyValuePair<string, Package>>())
+        {
+            // Ignore "file:" as these are local packages.
+            // Such local packages should only be referenced at the top level (via ProcessDependencyList) which also skips them or from other local packages (which this skips).
+            // There should be no cases where a non-local package references a local package, so skipping them here should not result in failed lookups below when adding all the graph references.
+            if (pnpmDependencyPath.StartsWith("file:"))
+            {
+                continue;
+            }
+
+            var parentDetectedComponent = PnpmParsingUtilities.CreateDetectedComponentFromPnpmPathV6(pnpmDependencyPath: pnpmDependencyPath);
+            components.Add(pnpmDependencyPath, (parentDetectedComponent, package));
+
+            // Register the component.
+            // It should get registered again with with additional information (what depended on it) later,
+            // but registering it now ensures nothing is missed due to a limitation in dependency traversal
+            // like skipping local dependencies which might have transitively depended on this.
+            singleFileComponentRecorder.RegisterUsage(parentDetectedComponent, isDevelopmentDependency: PnpmParsingUtilities.IsPnpmPackageDevDependency(package));
+        }
+
+        // Now that the `components` dictionary is populated, make a second pass registering all the dependency edges in the graph.
+        foreach (var (_, (component, package)) in components)
+        {
+            foreach (var (name, version) in package.dependencies ?? Enumerable.Empty<KeyValuePair<string, string>>())
+            {
+                var pnpmDependencyPath = PnpmParsingUtilities.ReconstructPnpmDependencyPathV6(name, version);
+
+                // If this lookup fails, then pnpmDependencyPath was either parsed incorrectly or constructed incorrectly.
+                var (referenced, _) = components[pnpmDependencyPath];
+
+                singleFileComponentRecorder.RegisterUsage(referenced, parentComponentId: component.Component.Id, isExplicitReferencedDependency: false);
+            }
+        }
+
+        // Lastly, add all direct dependencies of the current file/project setting isExplicitReferencedDependency to true:
+
+        // "dedicated shrinkwrap" (single package) case:
+        ProcessDependencySet(yaml);
+
+        // "shared shrinkwrap" (workspace / mono-repos) case:
+        foreach (var (_, package) in yaml.importers ?? Enumerable.Empty<KeyValuePair<string, PnpmHasDependenciesV6>>())
+        {
+            ProcessDependencySet(package);
+        }
+
+        void ProcessDependencySet(PnpmHasDependenciesV6 item)
```

Comment:
nit: can these be declared as explicit functions rather than locals for RecordDependencyGraphFromFileV6?

Reply:
I choose to do it this way so they could be scoped to v6 specific code, and not require as many parameters. They could of course be moved to a larger scope, given more specific names, and have everything passed in: that just seemed messier to me. If you refactor the v5 and v6 detector to be more separated, moving these to a larger scope would be more attractive.